flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/InputProcessorUtil.java

```diff
@@ -93,6 +97,12 @@ public static CheckpointedInputGate createCheckpointedInputGate(
 			toNotifyOnCheckpoint);
 		registerCheckpointMetrics(taskIOMetricGroup, barrierHandler);
 
+		if (config.isUnalignedCheckpointsEnabled()) {
```

Comment:
Maybe extract a common method `registerBufferReceivedListener(StreamConfig config, BufferReceivedListener listener, InputGate ...inputGates)` for duplicating two places.

Reply:
Not sure for N-array input processor, it is not a big issue for me. It is up to your judgement.